### ISSUE TITLE
Feature request: more transports (was: stdin/stdout transport for IPC)

### ISSUE THREAD

**Issue by sbstp** [google/tarpc#38]
The spawner process would act as the server and its children as clients. The child process's stdin and stdout can be treated like a socket. The child process could initiate the connection in its main.

I would use that to build a plugin system that allows live load/reload/unload-ing the plugins.

**Comment by tikue** [comment]
I'll take a crack at this. It'll probably end up on the async branch, which will get merged into master soon (tm).

**Comment by tikue** [comment]
To the async branch (tikue/async) I added `Client::new(Stream)`, as well as `ServeHandle::accept(Stream)` where

```enum Stream {
    Tcp(TcpStream),
    Pipe(PipeWriter, PipeReader),
}```

Let me know if you're able to get it working with that. I haven't tested it at all yet.

**Comment by tikue** [comment]
Added an [example](https://github.com/tikue/tarpc/blob/async/examples/pipe.rs) demonstrating the usage of pipes.

**Comment by sbstp** [comment]
Hey, thanks a lot for the quick reply. This looks good.

One thing I'm unsure about, is how to transform a `ChildStdin` into a `PipeWriter`.
For instance, the child process would try to connect to the master process in its main.
```
fn main() {
     let client = Client::connect((io::stdin(), io::stdout()));
}
```
But it might be impossible to do with mio. Looking quickly at the `Stdin`/`Stdout` structs, it doesn't look like they can be set non-blocking.

**Comment by tikue** [comment]
I spent a bit of time trying to figure that out last night too. I think the best thing to do would be to file an issue on either the rust or mio repo. I'm kind of the middle man here unfortunately.

**Comment by tikue** [comment]
@sbstp looks like mio [added](https://github.com/carllerche/mio/commit/f4aa49a9d2c4507fb33a4533d5238e0365f67c99) it a couple weeks ago.

**Comment by sbstp** [comment]
I've searched very briefly, and it doesn't look like there's an easy, cross platform way of configuring stdio to be unblocking. It behaves a lot like file I/O.

We could write a polyfill using background thread to simulate asynchronous I/O, or abandon the entire idea of using stdio.

**Comment by tikue** [comment]
@sbstp what about mio's support of it that I just linked?

**Comment by sbstp** [comment]
Looks like I didn't search properly. You can fnctl it like any other file descriptor. That fixes our issues then! Only need a new version of mio.

**Comment by tikue** [comment]
@sbstp yeah, I just added `#[cfg(unix)]` to tarpc's pipe and unix socket support.

**Comment by tikue** [comment]
This feature was lost in the tokio rewrite. I'm not sure how I feel about it now. I do like the simplicity of dealing strictly with TCP, but I realize that there's an opportunity here to be useful to a wider audience. I think I'll continue to leave this issue open, but I don't intend to implement it unless there's significant demand.

**Comment by tikue** [comment]
Should tarpc be generic over the transport layer (basically `Io`, or the new `AsyncRead + AsyncWrite`)? It seems worth considering, though I don't know if the hit to ergonomics is worth it. All clients and servers taking a type parameter seems rough. Maybe it's better to just extend StreamType to support a few more types of streams.

**Comment by tikue** [comment]
Obsoleted by #199. tarpc is now transport-agnostic.
